### PR TITLE
fix:ensure save row button is disabled when date column is required

### PR DESCRIPTION
**Pull request by saiprabhu-dandanayak**
### Bug : [Table date column bug](https://github.com/appsmithorg/appsmith/issues/34258)

I have raised this PR `In-order to ensure that the save row button is disabled when the date column is set to required`

### when department and doj are set to required save row is disabled
image

### when department has value and doj doesn't have , still the save row button is enabled
image

### when department and doj are set to required save row button is disabled and doj column indicating error
image

### when department has value and doj doesn't have , still the save row button is disabled
image

### when both the fieilds have values save row button is enabled
image

###  📸 Cypress Testing Video

https://github.com/zemoso-int/appsmith-from-the-business/assets/122865888/ba74fb1d-bd31-4549-a15a-6efdc01ed909

> [!CAUTION]  
> If you modify the content in this section, you are likely to disrupt the CI result for your PR.

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
  - Added date validation for "date" type columns in the table widget.
  
- **Bug Fixes**
  - Improved date formatting in date picker cells.
  
- **Improvements**
  - Simplified page size calculation for better performance.

### CODE REVIEW

**Comment by saiprabhu-dandanayak** [comment]
> **Actionable comments posted: 0**
> 
> Review details

Added a comment on the issue to review it.

**Comment by rahulbarwal** [comment]
@saiprabhu-dandanayak  The assets in PR description are not loading for me. Can you please re-upload the images and video?

**Comment by saiprabhu-dandanayak** [comment]
> @saiprabhu-dandanayak The assets in PR description are not loading for me. Can you please re-upload the images and video?

Hi **@rahulbarwal** , i have attached the same 
image
could you pls , check them once.

File: app/client/src/widgets/TableWidgetV2/component/cellComponents/DateCell.tsx [hunk 1]
```diff
@@ -241,8 +250,9 @@ export const DateCell = (props: DateComponentProps) => {
   };
 
   const onPopoverClosed = () => {
+    const isValidValue = value !== undefined && value !== null;
```
Comment (rahulbarwal) [inline app/client/src/widgets/TableWidgetV2/component/cellComponents/DateCell.tsx:253]:
```suggestion
    const isValidValue = !!value;
```
Also Don't we have to consider `isRequired` here?

**Review by rahulbarwal — CHANGES_REQUESTED** [review]
Functionally seems fine. Can you please also add some tests for these change?

File: app/client/src/widgets/TableWidgetV2/widget/derived.js [hunk 2]
```diff
@@ -187,9 +187,7 @@ export default {
         tableSizes.COLUMN_HEADER_HEIGHT) /
       tableSizes.ROW_HEIGHT;
 
-    return pageSize % 1 > 0.3 && props.tableData.length > pageSize
-      ? Math.ceil(pageSize)
-      : Math.floor(pageSize);
+    return pageSize % 1 > 0.3 ? Math.ceil(pageSize) : Math.floor(pageSize);
```
[the remark this answers is not in the document]
Comment (saiprabhu-dandanayak) [reply app/client/src/widgets/TableWidgetV2/widget/derived.js:190]: Hi  **@rahulbarwal** , that was not part of my issue , these were the changes suggested by coderabbitai

File: app/client/src/widgets/TableWidgetV2/component/cellComponents/DateCell.tsx (hunk 1 above)
Comment (saiprabhu-dandanayak) [reply app/client/src/widgets/TableWidgetV2/component/cellComponents/DateCell.tsx:253]: we have to considering `isRequired` also

**Comment by saiprabhu-dandanayak** [comment]
> Functionally seems fine. Can you please also add some tests for these change?

> Hi **@rahulbarwal** , will update the pr , with tests , for the current changes done.

File: app/client/src/widgets/TableWidgetV2/component/cellComponents/DateCell.tsx (hunk 1 above)
Comment (saiprabhu-dandanayak) [reply app/client/src/widgets/TableWidgetV2/component/cellComponents/DateCell.tsx:253]: Hi **@rahulbarwal** , will look into this and update the pr with expected results.

**Comment by rahulbarwal** [comment]
@saiprabhu-dandanayak any updates here?

**Comment by saiprabhu-dandanayak** [comment]
Working on testcases @rahulbarwal

**Comment by rahulbarwal** [comment]
@saiprabhu-dandanayak 
Hey checking up on this again. Are you still working on this? or can we take this ahead on our own. 
If you are stuck, can we assist you with something? 
cc: @jacquesikot

**Comment by saiprabhu-dandanayak** [comment]
> @saiprabhu-dandanayak Hey checking up on this again. Are you still working on this? or can we take this ahead on our own. If you are stuck, can we assist you with something? cc: @jacquesikot

Hi @rahulbarwal , it is working fine for new row , but not for existing one , i am working on it.

**Comment by jacquesikot** [comment]
@saiprabhu-dandanayak Is there anyway I can help with this? Are there any specific areas in the code you are struggling with?

**Comment by saiprabhu-dandanayak** [comment]
Hi **@jacquesiko**t , when we add `date`  in `validatedcolumns`  i.e [here](https://github.com/appsmithorg/appsmith/blob/7adb33d01590d9471fa79fe719026f64f64ea620/app/client/src/widgets/TableWidgetV2/widget/derived.js#L854) and adding relevant switch case , for newly added row `save row` button is being disabled but , for the existing rows , we are unable to add a date.

**Comment by jacquesikot** [comment]
@saiprabhu-dandanayak I will look into this today. Thanks

**Comment by jacquesikot** [comment]
Hello @saiprabhu-dandanayak I don't understand why we need to make changes in the derived.js file. I have made one change in the DateCell.tsx component, and it seems to solve all the problems reported without breaking the existing logic for existing cell values. I added

```
  useEffect(() => {
    if (isNewRow && isRequired && !value) {
      setIsValid(false);
    } else {
      setIsValid(true);
    }
  }, [value, isRequired, isNewRow]);
```

No other change was required, check this branch when you get the chance - [fix/disable-new-row-when-required-table-date-validation](https://github.com/appsmithorg/appsmith/tree/fix/disable-new-row-when-required-table-date-validation)

Let me know if this solves your problems. If it does, you can go ahead with tests and prop the final PR for review. Thanks again for the effort.

**Comment by saiprabhu-dandanayak** [comment]
Hi @jacquesikot , i have added `useEffect` in datecell component to check empty values and added `date` to validatedcells in derived.js file.

**Comment by jacquesikot** [comment]
@saiprabhu-dandanayak Can you check my branch and see if my chang solves the issues?

**Comment by saiprabhu-dandanayak** [comment]
Hi @jacquesikot , thanks for your support and will checkout to your branch and check functionality and move ahead with testcases , can you pls confirm check [#29662])https://github.com/appsmithorg/appsmith/issues/29662) this issue , i am  not able to reproduce. if it is already fixed pls close the issue.
Thank you.

**Comment by jacquesikot** [comment]
@saiprabhu-dandanayak To the best of my knowledge, the issues has not been fixed. I will try to reproduce on my end and send a loom video if i can.

**Comment by jacquesikot** [comment]
@saiprabhu-dandanayak I am also not able to reproduce, I will check the status of this and update you. Thanks

**Comment by saiprabhu-dandanayak** [comment]
Ok @jacquesikot

**Comment by jacquesikot** [comment]
Hey @saiprabhu-dandanayak any update? need any more help?

**Comment by saiprabhu-dandanayak** [comment]
Hi **@jacquesikot** , after incorporating your chnages the issue still not yet resolved , below is the screenshot for your reference.

### Screenshot
Screenshot from 2024-08-06 22-22-14

**Comment by jacquesikot** [comment]
@saiprabhu-dandanayak Did you check my branch to be sure it works appropriately there? The change I recommended is a standalone change, it should not be included with any other changes

**Comment by jacquesikot** [comment]
@saiprabhu-dandanayak Please review my branch again, I misunderstood your requirements, I have made some changes to the derived.js file within that branch, check if this solves your problems completely.

**Comment by saiprabhu-dandanayak** [comment]
Ok @jacquesikot , will check and let you know

**Comment by jacquesikot** [comment]
@saiprabhu-dandanayak Due to our priorities, I will be picking this issue up, I will update the PR i sent with tests and you can review when you get the chance

**Comment by jacquesikot** [comment]
@saiprabhu-dandanayak Thanks for helping with this, here is the merged PR that fixed the issue - https://github.com/appsmithorg/appsmith/pull/35613

**Pull request closed without merging**